flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/plan/batch/sql/SinkTest.scala

```diff
@@ -60,4 +67,25 @@ class SinkTest extends TableTestBase {
     util.verifyPlan()
   }
 
+  @Test
+  def testCatalogTableSink(): Unit = {
```

Comment:
actually, I find that the following simplified test will throw `NoMatchingTableFactoryException` before this PR, so I think this test case can be simplified.
```
  @Test
  def testCatalogTableSink(): Unit = {
    val schemaBuilder = new TableSchema.Builder()
    schemaBuilder.fields(Array("i"), Array(DataTypes.INT()))
    val schema = schemaBuilder.build()
    val catalog = util.tableEnv.getCatalog(util.tableEnv.getCurrentCatalog).get()
    val catalogTable = new CatalogTableImpl(schema, Map[String, String]().asJava, "")
    catalog.createTable(new ObjectPath("default_database", "tbl"), catalogTable, false)
    util.tableEnv.sqlQuery("select 1").insertInto("tbl")
    util.tableEnv.explain(false)
  }
```

Reply:
It seems this test case fails even with this PR. Because `GenericInMemoryCatalog::getTableFactory` returns Optional.empty(). So it falls back to the service discovery. Could you please verify on your side?

Reply:
oops.., i'm wrong before, i debug the case and stop at line 253 (`if (sink.isPresent) {`), i thought it could be succeed. 😂😂 

could you update `Mockito.atLeast(1)` to `Mockito.atLeast(2)` ? the first time is `    Mockito.when[Optional[_]](catalog.getTableFactory).thenReturn(Optional.of(factory))
` and another is in `getTableSink` method.

Reply:
I did a little bit test and seems `Mockito.when(...)` doesn't count as an invocation. The `getTableFactory` is called twice because we have `catalog.get().getTableFactory.isPresent` in `PlannerBase.getTableSink` and `catalog.getTableFactory().orElse(null)` in `TableFactoryUtil.createTableSinkForCatalogTable`. From a UT's perspective, these are implementation details. So I prefer not to verify "atLeast(2)". Does that make sense to you?